DDF-1960 works off comments on catalog/content merge

stustison opened the pull request:
#### What does this PR do?
Works off comments on catalog/content merge, fixes a couple of bugs
#### Who is reviewing it?
@pklinef @lessarderic 
#### How should this be tested?
unit/itests are sufficient
#### Any background context you want to provide?
#### What are the relevant tickets?
#### Screenshots (if appropriate)
#### Checklist:
- [ ] Documentation Updated
- [x] Update / Add Unit Tests
- [x] Update / Add Integration Tests

 - updated the content api with base request and response interfaces
 - updated the content impls
 - updated the content resource reader
 - refactored commit method in storage provider impl
 - fixed storage provider rollback
 - added unit tests
 - added integration tests

jlcsmith commented:
Care to fix this while you are at it? https://codice.atlassian.net/browse/DDF-1995

stustison commented:
Done.

coyotesqrl commented on catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/StorageProviderImpl.java:
```diff
@@ -127,44 +132,44 @@ public ReadStorageResponse read(ReadStorageRequest readRequest) throws StorageEx
+        try {
+            size = Files.size(file);
+        } catch (IOException e) {
+            LOGGER.warn("Unable to retrieve size of file.", e);
```
Since this means we'll get extra WARN logs when the file is accessed (from `StorageProviderImpl` when it checks this size against the actual data), we might want to include the name of the file in the log so those messages can be correlated.

coyotesqrl commented on catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/StorageProviderImpl.java:
```diff
@@ -235,110 +240,128 @@ public DeleteStorageResponse delete(DeleteStorageRequest deleteRequest)
+                    if (part1Files.size() == 0) {
+                        Files.deleteIfExists(part1);
                     }
+                    Path part0 = part1.getParent();
```
I think the delete of `part0` can be moved into the conditional on lines 281-283. If you aren't going to delete `part1`, it makes no sense to even check to delete its parent.

coyotesqrl commented:
There is definitely a bug in this PR. It's light-blue.

LGTM

lessarderic commented:
Some "left-overs" from the previous PR that haven't been addressed.

**`CatalogFrameworkImpl.create(CreateStorageRequest streamCreateRequest)`**

1. L877: Can’t we just check if `streamCreateRequest.getContentItems().size() == 0` at the top of the method and return right away if that’s the case? That would get rid of a bunch of checks later on in the method, make the code clearer and reduce cyclomatic complexity.
2. L917, 928: I believe these `if (createStorageRequest != null)` won't be needed anymore if comment #1 is addressed.
3. L920, 931, 936: Should we add `RuntimeException` to those catch blocks?
4. L921, 937: Pass exception to logging call. And should those be errors or warnings?

**`CatalogFrameworkImpl.guessFileExtension()`**
1. Can we rename  to something like `updateFileExtension()` or `getFileNameWithMimeTypeExtension()`? Would make it clearer that the method actually returns the full file name, not just the file extension.
2. L741: It's usually bad practice to reassign to a parameter

**`CatalogFrameworkImpl.guessMimeType()`**
1. L760, 770: Reassignment to a parameter

**`CatalogFrameworkImpl.update(UpdateStorageRequest)`**
1. There is a lot of duplication between this method and `create(CreateStorageRequest)` and most of the comments I made on the former apply here as well.
2. L1160: We should be able to make that check at the top and just return right away. Actually, looks like `validateUpdateStorageRequest()` is already checking for this, so we should be able to get rid of this check (same in `create(CreateStorageRequest)` method).
3. L1193: Do we care about duplicate IDs? Should we log something if we find duplicates?
4. L1209: Rename exception variable to `e` instead of `e1` and pass as argument to `LOGGER.error()`. Should we also add `RuntimeException` to the catch block? Should this be an error or a warning?
5. L1220: Same questions

**`CatalogFrameworkImpl.delete(DeleteRequest)`**
1. L1489: Pass `e` to `InternalIngestException` ctor. Also, shouldn't we log an error instead of a warning?

**`CatalogFrameworkImpl`**
1. L2052, 2070: Can we build these strings using `String.format()` instead?

pklinef commented:
Something I missed in the original PR is that the ContentItem interface could be immutable except for one setter (`setUri`).  I think we should remove that and only provider it on the impl.

https://github.com/codice/ddf/blob/885e04dc9653a5dac75aec7e992ee0a8e4692968/catalog/core/catalog-core-api/src/main/java/ddf/catalog/content/data/ContentItem.java#L68

lessarderic commented on catalog/core/catalog-core-api/src/main/java/ddf/catalog/content/operation/StorageRequest.java:
```diff
@@ -0,0 +1,33 @@
+     * {@link ddf.catalog.content.StorageProvider} to {@link ddf.catalog.content.StorageProvider#commit(StorageRequest)}
+     * or {@link ddf.catalog.content.StorageProvider#rollback(StorageRequest)} changes.
+     *
+     * @return ID
```
Since those are brand new APIs, could we indicate in the comments what could be `null` and what won't. That helps clients as well as implementers. Not going to ask you to go as far as adding `@NotNull` on return values and arguments, but I won't complain if you do ;-)

stustison replied:
How about I just comment on things that _can_ be null, since I don't like using nulls anyways and I don't feel like adding a comment to every single parameter.

lessarderic replied:
We need to come up with some commenting standards for our new APIs. From a client perspective, _not_ saying anything on a method (especially on returned values) is ambiguous and if you want to play safe, you _have_ to check for `null`, which is just bad.

lessarderic commented on catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/content/data/impl/ContentItemImpl.java:
```diff
@@ -32,30 +32,32 @@
      * <code>null</code> because the {@link ddf.catalog.CatalogFramework} will assign its GUID.
      *
-     * @param byteSource          the {@link ContentItem}'s input stream containing its actual data
+     * @param byteSource      the {@link ContentItem}'s input stream containing its actual data
```
Are all the arguments required (non-`null`)? Same question for the other ctor.

stustison replied:
See above.

lessarderic commented on catalog/core/catalog-core-api/src/main/java/ddf/catalog/content/operation/CreateStorageRequest.java:
```diff
@@ -25,7 +24,7 @@
+public interface CreateStorageRequest extends StorageRequest {
 
     /**
      * Returns {@link List} of {@link ContentItem}
```
Instead of repeating the same thing here, should we just keep the `@return`? We should also specify if the list returned by `getContentItems()` can be `null`.

lessarderic commented:
The `CreateStorageRequestImpl` and `UpdateStorageRequestImpl` classes basically contain the same code. Would it make sense to have a common base class?

Also, can we remove the `setContentItem()` on those classes and make the classes immutable? If we need to keep it, we probably should check that the parameter is not `null` like we do in the constructor.

lessarderic commented:
Do we really need 4 different `StorageResponseImpl` classes? Except for `ReadStorageResponseImpl`, they're all pretty much the same.

lessarderic commented on catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/StorageProviderImpl.java:
```diff
@@ -51,6 +53,9 @@
+/**
+ * File system storage provider.
+ */
 public class StorageProviderImpl implements StorageProvider {
```
How about `FileSystemStorageProvider` instead of `StorageProviderImpl`?

stustison replied:
OK.

lessarderic commented on catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/StorageProviderImpl.java:
```diff
@@ -235,110 +240,128 @@ public DeleteStorageResponse delete(DeleteStorageRequest deleteRequest)
 
     @Override
-    public void commit(String id) throws StorageException {
+    public synchronized void commit(StorageRequest request) throws StorageException {
```
The current synchronization approach won't make this class thread-safe. We should either synchronize on the `deletionMap` field (my preference) or synchronize all the methods that access the map.

stustison replied:
Forgot to synchronize the delete method, added.

lessarderic commented on catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/StorageProviderImpl.java:
```diff
@@ -235,110 +240,128 @@ public DeleteStorageResponse delete(DeleteStorageRequest deleteRequest)
-                rollback(id);
             }
+        } finally {
+            rollback(request);
```
I understand that we need to rollback on failure, but why do we rollback on success?

lessarderic commented on catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/StorageProviderImpl.java:
```diff
@@ -235,110 +240,128 @@ public DeleteStorageResponse delete(DeleteStorageRequest deleteRequest)
+                }
+            }
+        } finally {
+            rollback(request);
```
I understand that we need to rollback on failure, but why do we rollback on success?

stustison replied:
This just ensures that the temp directory is clean after a commit.

lessarderic commented on catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/StorageProviderImpl.java:
```diff
@@ -235,110 +240,128 @@ public DeleteStorageResponse delete(DeleteStorageRequest deleteRequest)
 
     @Override
-    public void rollback(String id) throws StorageException {
+    public synchronized void rollback(StorageRequest request) throws StorageException {
```
This method looks a lot more like a clean up method than a rollback. Shouldn't a rollback delete the files that were created from the store, restore the previous version of the files that were updated and restore the files that were deleted?

stustison replied:
That is essentially what it is doing. You're looking too far into my implementation of _how_ I'm doing that as opposed to looking at what is occurring. When rollback is called, the request is essentially reverted, whether it was a create, update, or delete.

lessarderic commented:
**`StorageProviderImpl`**
1. L115: Should we log something here?
2. L146: Can we catch specific checked exceptions and `RuntimeException` here instead of `Exception`?
3. We should consider extracting the code that deals with the directory structure to a different class. Other than the usual advantages (smaller classes that are easier to understand, code easier to unit test, etc.), we could make the extracted class a little more generic and potentially reusable (I think the `dump` and `export` commands actually do something similar). We could also make the implementation vary in the future if need be. For instance, we could have the current implementation that uses the first 6 characters of the metacard ID to distribute the files between directories, but be able to substitute the implementation to something that distributes them some other way based on how the IDs are formatted.
4. L475, 479: Use existing constant for `karaf.home` string or create one in this class.

stustison commented:
Re: I missed a couple of these and others I didn't feel needed to be completed. I'll reply inline.
Some "left-overs" from the previous PR that haven't been addressed.

CatalogFrameworkImpl.create(CreateStorageRequest streamCreateRequest)

L877: Can’t we just check if streamCreateRequest.getContentItems().size() == 0 at the top of the method and return right away if that’s the case? That would get rid of a bunch of checks later on in the method, make the code clearer and reduce cyclomatic complexity.
`I'm not sure why this would be needed. If there are no content items, the method exits immediately after checking if the request is local anyways`
L917, 928: I believe these if (createStorageRequest != null) won't be needed anymore if comment #1 is addressed.
`see above`
L920, 931, 936: Should we add RuntimeException to those catch blocks?
`I don't think so, if we encounter a runtime exception anywhere in here, that would be bad.`
L921, 937: Pass exception to logging call. And should those be errors or warnings?
`missed this one, updated`
CatalogFrameworkImpl.guessFileExtension()
1. Can we rename to something like updateFileExtension() or getFileNameWithMimeTypeExtension()? Would make it clearer that the method actually returns the full file name, not just the file extension.
`updated`
2. L741: It's usually bad practice to reassign to a parameter
`I generally agree. I usually make an exception when the method is only working on the parameter that I'd be reassigning anyways. I would essentially just be assigning it to a different variable at the top and never touching the parameter from that point on. That just seems like a wasted line imo.`

CatalogFrameworkImpl.guessMimeType()
1. L760, 770: Reassignment to a parameter
`see above`

CatalogFrameworkImpl.update(UpdateStorageRequest)
1. There is a lot of duplication between this method and create(CreateStorageRequest) and most of the comments I made on the former apply here as well.
`I removed a few more duplicate lines.`
2. L1160: We should be able to make that check at the top and just return right away. Actually, looks like validateUpdateStorageRequest() is already checking for this, so we should be able to get rid of this check (same in create(CreateStorageRequest) method).
`different list of content items`
3. L1193: Do we care about duplicate IDs? Should we log something if we find duplicates?
`duplicates are fine`
4. L1209: Rename exception variable to e instead of e1 and pass as argument to LOGGER.error(). Should we also add RuntimeException to the catch block? Should this be an error or a warning?
5. L1220: Same questions

CatalogFrameworkImpl.delete(DeleteRequest)
1. L1489: Pass e to InternalIngestException ctor. Also, shouldn't we log an error instead of a warning?
`updated`

CatalogFrameworkImpl
1. L2052, 2070: Can we build these strings using String.format() instead?
`maybe, but I hate string.format`

stustison commented:
Re:
StorageProviderImpl
1. L115: Should we log something here?
`not if it is throwing`
2. L146: Can we catch specific checked exceptions and RuntimeException here instead of Exception?
`what would be the difference, besides having more exceptions called out in the catch?`
3. We should consider extracting the code that deals with the directory structure to a different class. 
`I might be missing what you're referring to. The code that deals with the directory structure is all of 10 lines long.`
Other than the usual advantages (smaller classes that are easier to understand, code easier to unit test, etc.), we could make the extracted class a little more generic and potentially reusable (I think the dump and export commands actually do something similar). We could also make the implementation vary in the future if need be. For instance, we could have the current implementation that uses the first 6 characters of the metacard ID to distribute the files between directories, but be able to substitute the implementation to something that distributes them some other way based on how the IDs are formatted.
4. L475, 479: Use existing constant for karaf.home string or create one in this class.
`done`
